### PR TITLE
Alert rules

### PR DESCRIPTION
**Pull request by jr1221**
## Changes

_Explanation of changes goes here_

## Notes

_Any other notes go here_

## Test Cases

- Case A
- Edge case
- ...

## Screenshots

_If you made UI changes you must post a screenshot of the whole page for each change in 1) a normal sized window and 2) the smallest possible window_

_If you did any manual testing (e.g., with Postman), put screenshots of the http request and before and after of the db_

_If none of this applies, you can delete this section_

## To Do

_Any remaining things that need to get done_

- [ ] item 1
- [ ] ...

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://nerdocs.atlassian.net/wiki/spaces/NER/pages/8060929/Software+Contributor+Guide) and reach out to your Tech Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [ ] All commits are tagged with the ticket number
- [ ] No linting errors / newline at end of file warnings
- [ ] All code follows repository-configured prettier formatting
- [ ] No merge conflicts
- [ ] All checks passing
- [ ] Screenshots of UI changes (see Screenshots section)
- [ ] Remove any non-applicable sections of this template
- [ ] Assign the PR to yourself
- [ ] No `package-lock.json` changes (unless dependencies have changed)
- [ ] Request reviewers & ping on Slack
- [ ] PR is linked to the ticket (fill in the closes line below)

Closes # (issue #)

### CODE REVIEW

File: scylla-server/src/controllers/rule_controller.rs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+) -> Result<Json<String>, ScyllaError> {
+    debug!(
+        "Incoming rules reg: {}, from {}",
+        rule.topic,
+        auth.username().to_string()
+    );
+    match rules_manager
+        .write()
+        .await
+        .add_rule(ClientId(auth.username().to_string()), rule)
```
Comment (bracyw) [inline scylla-server/src/controllers/rule_controller.rs:30]: might answer this myself later: ClientID... usernames ... how do they work?

File: scylla-server/src/rule_structs.rs [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+    /// process a tick (seeing this topic)
+    /// returns a boolean of whether the rule was triggered, or None if err
+    pub fn tick(&mut self, values: &[f32]) -> Option<bool> {
+        if self.during_cooldown {
+            // check cooldown should still be happening then bail if so
+            let Some(last_seen) = self.last_seen else {
+                warn!("Don't know when cooldown began!");
+                return None;
+            };
+            if tokio::time::Instant::now() - last_seen < COOLDOWN_TIME {
```
Comment (bracyw) [inline scylla-server/src/rule_structs.rs:117]: could cool down be configurable... might be something we want in the future?

File: scylla-server/src/rule_structs.rs [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+            rules_lookup: FxHashMap::default(),
+        }
+    }
+
+    /// Handles a new socket message, returning a RuleNotification for one to many clients if action should be taken
+    pub fn handle_msg(
+        &mut self,
+        data: &ClientData,
+    ) -> Result<Option<Vec<(ClientId, RuleNotification)>>, RuleManagerError> {
+        // TODO uneccessary clone
```
Comment (bracyw) [inline scylla-server/src/rule_structs.rs:182]: still todo... or is this a laterrrrr todo?

File: scylla-server/src/socket_handler.rs [hunk 4]
```diff
@@ -142,13 +192,47 @@ pub async fn socket_handler_with_metadata(
     }
 }
 
+async fn handle_rule_stuff(
```
Comment (bracyw) [inline scylla-server/src/socket_handler.rs]: maybe add a comment / rename

File: scylla-server/src/socket_handler.rs [hunk 5]
```diff
@@ -69,6 +82,41 @@ pub async fn socket_handler_with_metadata(
         Regex::new(r"MPU\/Fault\/Critical\/(.*)").expect("Could not compile regex!");
     let mut fault_ringbuffer = AllocRingBuffer::<FaultData>::new(25);
 
+    // END METADATA
+
+    // BEGIN rules
+    // socket_map must not be written outside of these closures.
+    // a map of client_ids to their respective socket IDs for identifying where notifications go
+    let client_socket_map: Arc<RwLock<FxHashMap<String, Sid>>> =
```
Comment (bracyw) [inline scylla-server/src/socket_handler.rs:90]: thinking it will work better long term to move this collection of client_ids and socket_ids into a separate function. So that way if we need to we can give access to the client_socket_map outside of the handle_with_metedata function.

File: scylla-server/src/rule_structs.rs [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+        }
+
+        for client_want in clients {
+            let Some(rules) = self.clients_map.get_mut(client_want) else {
+                warn!("Client cached but not found!");
+                return Err(RuleManagerError::Failure);
+            };
+
+            for rule in rules.values_mut() {
+                if rule.topic == topic {
```
Comment (bracyw) [inline scylla-server/src/rule_structs.rs]: I'm thinking this will be expensive if we even have one person with a lot of rules and more topics per second. Maybe  make the rules a HashSet of `<topic, rules[]>` for a quick fix. Probably a better way to do it though.

**Review by bracyw — DISMISSED** [review]
Couple questions, and maybe changes to look up of rules on each message

File: scylla-server/src/controllers/rule_controller.rs (hunk 1 above)
Comment (jr1221) [reply scylla-server/src/controllers/rule_controller.rs:30]: https://nerdocs.atlassian.net/wiki/spaces/NER/pages/970358819/Argos+Alert+Rules
Basically the username is the identifier that is given to the backend so we know which client wants what notifications via what rules.

File: scylla-server/src/rule_structs.rs (hunk 2 above)
Comment (jr1221) [reply scylla-server/src/rule_structs.rs:117]: Definitely.  I dont have time to add that rn but could be followup issue?

File: scylla-server/src/rule_structs.rs (hunk 3 above)
Comment (jr1221) [reply scylla-server/src/rule_structs.rs:182]: well later is fine ig.

File: scylla-server/src/socket_handler.rs (hunk 5 above)
Comment (jr1221) [reply scylla-server/src/socket_handler.rs:90]: Can you elaborate?  Right now its just a variable local to handle_with_metadata.  I figured that wold be ok because this rules stuff only works with metadata enabled.

File: scylla-server/src/rule_structs.rs (hunk 6 above)
Comment (jr1221) [reply scylla-server/src/rule_structs.rs]: Smart.  Latest commit addresses that, we will see if it works lol as it is a restructure a little bit.

File: scylla-server/src/socket_handler.rs (hunk 5 above)
Comment (bracyw) [reply scylla-server/src/socket_handler.rs:90]: We can do it later if needed not important rn

**Review by bracyw — APPROVED** [review]
